accord-core/src/main/java/accord/local/CommandStore.java

```diff
@@ -534,30 +553,77 @@ public final void removeRedundantDependencies(Unseekables<?> participants, Waiti
             }
             return b;
         }, builder, keyDeps, null, ignore -> false);
-        RangeDeps rangeDeps = builder.deps.rangeDeps;
-        // TODO (now): slice to only those ranges we own, maybe don't even construct rangeDeps.covering()
-        redundantBefore().foldl(participants, (e, b, d, rs, pi, pj) -> {
-            // TODO (now): foldlInt so we can track the lower rangeidx bound and not revisit unnecessarily
-            boolean useBootstrap = e.isLocalRedundancyViaBootstrap();
+
+        class RangeState
+        {
+            final WaitingOn.Update builder;
             int txnIdx;
+            Range range;
+            Map<Integer, Ranges> partiallyBootstrapping;
+
+            RangeState(WaitingOn.Update builder)
+            {
+                this.builder = builder;
+            }
+
+            boolean isFullyBootstrapping(int rangeTxnIdx)
+            {
+                if (partiallyBootstrapping == null)
+                    partiallyBootstrapping = new HashMap<>();
+                Ranges prev = partiallyBootstrapping.get(rangeTxnIdx);
+                Ranges remaining = prev;
+                if (remaining == null) remaining = builder.deps.rangeDeps.ranges(rangeTxnIdx);
+                else Invariants.checkState(!remaining.isEmpty());
+                remaining = remaining.subtract(Ranges.of(range));
+                if (prev == null) Invariants.checkState(!remaining.isEmpty());
+                partiallyBootstrapping.put(txnIdx, remaining);
+                return remaining.isEmpty();
+            }
+        }
+        RangeDeps rangeDeps = builder.deps.rangeDeps;
+        // TODO (required, consider): slice to only those ranges we own, maybe don't even construct rangeDeps.covering()
+        redundantBefore().foldl(participants, (e, s, d, ps, pi, pj) -> {
+            // TODO (desired, efficiency): foldlInt so we can track the lower rangeidx bound and not revisit unnecessarily
+            WaitingOn.Update b = s.builder;
             {
-                int tmp = d.txnIds().find(useBootstrap ? e.bootstrappedAt : e.redundantBefore);
-                txnIdx = tmp < 0 ? -1 - tmp : tmp;
+                int tmp = d.txnIds().find(e.redundantBefore);
+                s.txnIdx = tmp < 0 ? -1 - tmp : tmp;
             }
-            rangeDeps.forEach(participants, e.range, pi, pj, d, (d0, txnIdx0) -> {
-                if (txnIdx0 < txnIdx)
-                    b.setAppliedOrInvalidatedRangeIdx(txnIdx0);
+            d.forEach(ps, e.range, pi, pj, b, s, (b0, s0, txnIdx) -> {
+                if (txnIdx < s0.txnIdx)
+                    b0.setAppliedOrInvalidatedRangeIdx(txnIdx);
             });
-            return b;
-        }, builder, rangeDeps, participants, ignore -> false);
+            if (e.bootstrappedAt.compareTo(e.redundantBefore) > 0)
```

Comment:
Redundant is very like bootstrap, but much simpler: if a transaction is redundant, then we have applied it locally and we don't care about it. If it is pre-bootstrap, it might still need to be partially applied, so we can only (but must) remove it if all of the keys on which we depend upon it are pre-bootstrap.